Source/Session.swift

```diff
@@ -465,18 +465,25 @@ open class Session {
         requestTaskMap[request] = task
         request.didCreateTask(task)
 
-        resumeOrSuspendTask(task, ifNecessaryForRequest: request)
+        updateStatesForTask(task, request: request)
     }
 
-    func resumeOrSuspendTask(_ task: URLSessionTask, ifNecessaryForRequest request: Request) {
-        if startRequestsImmediately || request.isResumed {
+    func updateStatesForTask(_ task: URLSessionTask, request: Request) {
+        switch (startRequestsImmediately, request.state) {
+        case (true, .initialized):
+            request.resume()
+        case (false, .initialized):
+            // Do nothing.
+            break
+        case (_, .resumed):
             task.resume()
-            request.didResume()
-        }
-
-        if request.isSuspended {
+            request.didResumeTask(task)
+        case (_, .suspended):
             task.suspend()
-            request.didSuspend()
+            request.didSuspendTask(task)
+        case (_, .cancelled):
+            task.cancel()
+            request.didCancelTask(task)
         }
```

Comment:
I think there is still a race here, but I don't think it ends up having any undesirable behavior from what I can tell. Stick with me here for a sec.

The `Request.resume`, `Request.suspend`, and `Request.cancel` APIs do not pipe through the `rootQueue` in any way. The transition logic runs inside the protector, but not inside the `rootQueue`. Therefore, the `Protector` seems to be the only thread-safety mechanism.

If this is the case (certainly possible it's not), then it's possible to have the following happen:

- `updateStatesForTask` is called on rootQueue and `request.state` if pulled from `Protecter`
- Another thread calls `Request.cancel()` and changes the state to cancelled
- The `updateStatesForTask` continues with a now out-of-date state (let's say it is initialized)
- The `request.resume()` method is called
- The attempt to transition to the `resumed` state will fail silently

If that's all correct, then there's still a race, but I don't think it has any side effects given all the `resume`, `suspend`, and `cancel` Request APIs all attempt to transition back through the protector and will be stopped if they can't transition.

I wanted to point this out 1) for my own understanding and 2) in case this was not your intent for this logic. Anyways, I think this is solid from what I can tell.

Reply:
You're correct: there's a race on `request.state`. Luckily, as you noticed, the result doesn't matter, as both code paths represent correct behavior. Obviously, changing the state before `updateStatesForTask(_:request:)` is called leads directly to the outcome we want. In your scenario, the change to `request.state` is immediate, but `rootQueue` is already executing inside `updateStatesForTask`'s `switch`. So the created `task` will have the action called, and `request` will receive the appropriate task callback. However, when the `Request` action was run, after `state` is changed, the `delegate` was called and its actions were enqueued on `rootQueue`. Once the queue runs that closure we just have to guarantee that the correct action are performed.

In summary, since we know the correct actions are always enqueued after `updateStatesForTasks`, we always behave correctly.

I've added a cancellation test for an automatically resumed `Request` being cancelled from many queues concurrently in https://github.com/Alamofire/Alamofire/pull/2796/commits/f354b4bfacbe1ffeaad99d5f0aa42277a2e7ec4c. We can add more, but this should give us coverage of the worst case scenario.

Reply:
Yep, good callout on the lazy task creation. That will certainly require some internal refactoring of this type of logic, but should still be possible without breaking public API changes.